ext/bg/js/translator.js

```diff
@@ -518,4 +556,39 @@ class Translator {
         const pos = name.indexOf(':');
         return (pos >= 0 ? name.substring(0, pos) : name);
     }
+
+    static *getArrayVariants(arrayVariants) {
+        const ii = arrayVariants.length;
+
+        let total = 1;
+        for (let i = 0; i < ii; ++i) {
+            total *= arrayVariants[i].length;
+        }
+
+        for (let a = 0; a < total; ++a) {
+            const variant = [];
+            let index = a;
+            for (let i = 0; i < ii; ++i) {
+                const entryVariants = arrayVariants[i];
+                variant.push(entryVariants[index % entryVariants.length]);
+                index = Math.floor(index / entryVariants.length);
+            }
+            yield variant;
+        }
+    }
+
+    static getSearchableText(text, options) {
+        if (!options.scanning.alphanumeric) {
+            const ii = text.length;
+            for (let i = 0; i < ii; ++i) {
+                const c = text[i];
+                if (!jpIsCharacterJapanese(c)) {
+                    text = text.substring(0, i);
+                    break;
+                }
+            }
+        }
+
+        return text;
+    }
```

Comment:
This is probably more correct than the previous approach of checking the first character only, but it uncovered a bug: in 散々 the kanji iteration mark isn't detected as Japanese text because `jpIsCharacterJapanese` only checks kanji and kana.

WanaKana has its [own version](https://wanakana.com/docs/global.html#isJapanese) that handles quite many special cases: https://github.com/WaniKani/WanaKana/blob/cf3f9e948d96ea40c311135b5082d1ffa77a96bf/src/constants.js#L102

Reply:
You're right; I also just noticed I introduced a bug with `jpIsKana` since it now only checks a single character:
https://github.com/FooSoft/yomichan/blob/a163a37811f0b230023553d0f1b371d29069902d/ext/mixed/js/japanese.js#L162
^ used to check the entire string.

I moved away from using WanaKana's implementations directly for these functions for two reasons:
* Yomichan's `jp*` character test functions are (for the most part) only testing a single character. With this assumption, the code can be optimized a bit.
* WanaKana's `isJapanese` implemention looks somewhat unoptimized. WanaKana is written for a general use case/environment, but we can make more assumptions based on Yomichan's more specific use cases. (No good way to reference the built code, but you can search for the function definition in the [non-minimized build file](https://unpkg.com/wanakana@4.0.2/umd/wanakana.js).)

Do you see any issue with copying some of the range definitions WanaKana uses and repurposing them inside `japanese.js`? If not, I'll probably do that and do a bit more cleanup and fixing of some of the other `jp*` character testing methods.

Another thing I realized: I don't think `japanese.js` should be in the `mixed` folder, since it's only used by the backend. Using it on the frontend would have an undefined `wanakana` reference.

Reply:
> Do you see any issue with copying some of the range definitions WanaKana uses and repurposing them inside japanese.js? If not, I'll probably do that and do a bit more cleanup and fixing of some of the other jp* character testing methods.

The only issue I see is having to maintain our own version of a Japanese range detector, which might get changes to its magic numbers or logic in the future when Unicode adds more characters or Japan decides to mix things up, or there's something that hasn't been thought of yet and has to be added. WanaKana (WaniKani) probably has more knowledge and preparation about that kind of stuff.

Anyway, I understand the point that our use case is quite different from what most people need (a browser based kana IME or simple range checker), and it might have a performance impact when you do checks as fast as your browser can emit `mousemove` events. So the priorities of this issue for me are:

1. Have the `jp*` functions abstracted and give them well-defined behavior such as checking a single character at once only
2. Keep scanning as fast as possible
3. If performance doesn't take a hit, prefer using an external lib to minimize the maintenance cost

> Another thing I realized: I don't think japanese.js should be in the mixed folder, since it's only used by the backend. Using it on the frontend would have an undefined wanakana reference.

Probably safe to move it to `bg`. I also grepped the function names in `japanese.js` and looks good.

Reply:
Another interesting note: WanaKana's `JAPANESE_RANGES` includes `KANA_PUNCTUATION` twice. The order of `JAPANESE_RANGES` is also sub-optimal for the general case since the kanji ranges are at the end.

> Probably safe to move it to bg. I also grepped the function names in japanese.js and looks good.

Cool, I'll probably do that after this commit because it's more of a housekeeping task.

Reply:
One other thing related to `japanese.js`: DisplayGenerator has its own set of ranges for kanji which was copied out because DisplayGenerator didn't have access to `japanese.js` because it can be used on pages which don't include it.

https://github.com/FooSoft/yomichan/blob/04727a8cd7dc29a02595690389979ec1070bbded/ext/mixed/js/display-generator.js#L326-L332

Reply:
> DisplayGenerator has its own set of ranges for kanji which was copied

In that case maybe it would still make sense to have `japanese.js` in mixed and add it to pages where `display-generator.js` is used, but that isn't many lines of duplication.